Andrew/spikev3

andrewjgremmo opened the pull request:
# Role Changes
- Add Spike Captain role
- Limit Spike Captain to viewing (not interacting w/) button

# Spiker Changes
- Add notification preference to spikers
  - Spike alert call/text/both based on preferences
- Add spiker signup form
- Notify Spike Captain when new Spiker signs up

# Spiker Group Changes
- Remove disabling of groups
- Remove shuffling of groups
- Convert spikers to days of the week model

# Button Changes
- Only display today's group and "all" options for button push
- Limit button to two pushes per day

# Misc Changes
- Remove stats dashboard

# Bug Fixes
- Fix #27

kidatsy commented on app/DoctrineMigrations/Version20160219100158.php:
```diff
@@ -0,0 +1,28 @@
+
+        $this->addSql('ALTER TABLE spiker DROP notification_preference');
+    }
+}
```
could we merge these two migrations?

andrewjgremmo replied:
Yeah we could do that, my logic here was breaking up migrations by feature so that one feature could be rolled up/down without effecting the other.  Unlikely to happen in prod so I could be convinced otherwise.

kidatsy replied:
that would prob be more of a concern if we were rolling them up into separate PRs/pushes, and i dont see us going back on either of these independently. so i dont think it would be bad to combine.

kidatsy commented on app/Resources/views/admin.html.twig:
```diff
@@ -18,11 +18,12 @@
+            {% if is_granted('ROLE_ADMIN') %}
+              <li class="admin"><a href="{{ path('admin') }}">Admin Users</a></li>
+              <li class="settings"><a href="{{ path('settings') }}">Settings</a></li>
+            {% endif %}
```
:facepalm:

kidatsy commented on app/config/parameters.yml.dist:
```diff
@@ -18,3 +18,6 @@ parameters:
     # A secret key that's used to generate certain security-related tokens
     secret:            ThisTokenIsNotSoSecretChangeIt
+
+    host:              127.0.0.1:8000
```
where is this being used?

andrewjgremmo replied:
This is used for the Twilio call stuff.  The callback needs to pass an absolute URL and I've had issues with generating absolute URLs outside of the controller context in the past.  Actually, now that you bring it up, this may be unnecessary because the last time I had to generate absolute URLs was from a cronjob so it didn't have any web context to pull the protocol and host from, I'll give it a shot the other way.

kidatsy replied:
sounds good

kidatsy commented on src/SpikeTeam/AlertBundle/Tests/Event/AlertListenerTest.php:
```diff
@@ -0,0 +1,85 @@
+            ->setMethods($methods)
+            ->getMock();
+    }
+}
```
nice

kidatsy commented on src/SpikeTeam/ButtonBundle/Controller/ButtonController.php:
```diff
@@ -9,27 +9,36 @@
+                     ->getRepository('SpikeTeamUserBundle:SpikerGroup')
+                     ->find($currentGroupId);
+
+        $canPush = $currentGroup->getRecentPushesCount() < 2;
```
set the push limit (2) as a setting?

andrewjgremmo replied:
Yeah that's a good idea

kidatsy commented on src/SpikeTeam/ButtonBundle/Controller/ButtonController.php:
```diff
@@ -63,7 +91,7 @@ public function goAction($gid)
             $em->flush();
 
-            $next = $this->getNextGroup();
+            $next = $spikerGroupHelper->getCurrentGroupId();
```
if we're no longer doing the funky group rotation thing, can we rename this to something that's not next, just to be explicit?

andrewjgremmo replied:
Good call

kidatsy commented on src/SpikeTeam/UserBundle/Controller/SpikerController.php:
```diff
@@ -44,45 +45,24 @@ public function spikersAllAction(Request $request, $group = null)
+        $form = $this->createForm(new SpikerType(), $newSpiker)
+                     ->remove('notificationPreference')
+                     ->remove('isCaptain')
+                     ->remove('isEnabled');
```
wow, did not know you could do that. awesome!

andrewjgremmo replied:
Me neither until this.  I wish there was a #modify method.  You'll see later I had to remove and then add a field just to modify it.  :(

kidatsy replied:
:)

kidatsy commented on src/SpikeTeam/UserBundle/Controller/SpikerController.php:
```diff
@@ -158,101 +174,42 @@ public function spikerEditAction($input, Request $request)
-                    $this->em->flush();
-                    return $this->redirect($allUrl);
-                } else {
-                    return $this->redirect($editUrl);
```
thanks for cleaning all this nonsense up. the joys of abstracted form types!

kidatsy commented on src/SpikeTeam/UserBundle/Entity/Spiker.php:
```diff
@@ -5,12 +5,22 @@
+ * @UniqueEntity(
+ *     fields="phoneNumber",
+ *     message="This phone number is already signed up."
+ * )
```
+1

kidatsy commented on src/SpikeTeam/UserBundle/Tests/Controller/SpikerGroupControllerTest.php:
```diff
@@ -6,50 +6,47 @@
-        // Check the entity has been delete on the list
-        $this->assertNotRegExp('/Foo/', $client->getResponse()->getContent());
+        // // Check the entity has been delete on the list
+        // $this->assertNotRegExp('/Foo/', $client->getResponse()->getContent());
```
test coming here?

kidatsy replied:
test still commented?

kidatsy commented on web/js/button.js:
```diff
@@ -13,23 +13,14 @@ var Button = {
-                        + "* (10pm EST till 6am EST of the following day)"
-                    ) == true) {
+                if (confirm("Are you sure you want to notify the " +
+                    days[groupId] + " Spike Team?")) {
```
can we get rid of the days array here and go directly with the group name, now that we're explicitly setting it in DB, rather than relying on id?

andrewjgremmo replied:
Oh, yeah, that makes more sense haha, derp.

kidatsy commented:
overall, looks good, aside from the few nits up there. will pull it in and play with it a bit. thanks for cleaning up all that muck!

kidatsy commented on app/config/security.yml:
```diff
@@ -37,7 +38,7 @@ security:
     access_control:
         - { path: ^/twilio, roles: IS_AUTHENTICATED_ANONYMOUSLY, requires_channel: http }
         - { path: ^/login$, roles: IS_AUTHENTICATED_ANONYMOUSLY, requires_channel: https }
```
trying to access spikers/signup as anon redirects to login. maybe need to open up that route in the firewall?

kidatsy commented on src/SpikeTeam/UserBundle/Controller/SpikerController.php:
```diff
@@ -150,6 +130,42 @@ public function spikerEnablerAction(Request $request)
+        // send to template
+        return $this->render('SpikeTeamUserBundle:Spiker:signup.html.twig', array(
+            'form' => $form->createView(),
+            'errors' => (isset($errors)) ? $errors : null
```
successful signup redirects to here:
![screen shot 2016-02-25 at 12 14 05 pm](https://cloud.githubusercontent.com/assets/2397451/13327905/dd6d0cc2-dbb9-11e5-9a47-fdf8103039a9.png)


can we redirect somewhere more.. friendly?

kidatsy commented on src/SpikeTeam/UserBundle/Resources/views/Admin/adminsAll.html.twig:
```diff
@@ -11,7 +12,7 @@
+                <th>Role</th>
                 <th>First Name</th>
                 <th>Last Name</th>
                 <th>Email</th>
```
nit: can we add th for phone here?
![screen shot 2016-02-25 at 12 14 37 pm](https://cloud.githubusercontent.com/assets/2397451/13327957/26af5048-dbba-11e5-9c76-b631790642a2.png)

andrewjgremmo replied:
Good catch

kidatsy commented on app/Resources/views/base.html.twig:
```diff
@@ -28,7 +28,6 @@
             <script src="{{ asset('js/app.js') }}"></script>
             <script src="{{ asset('js/button.js') }}"></script>
             <script src="{{ asset('js/groups.js') }}"></script>
-            <script src="{{ asset('js/stats.js') }}"></script>
```
we're going to have to move the js into the respective bundles eventually. meh.

kidatsy commented on web/js/button.js:
```diff
@@ -13,23 +13,14 @@ var Button = {
+                if (confirm("Are you sure you want to notify the " +
+                    days[groupId] + " Spike Team?")) {
                     $(this).removeClass("enabled").addClass("disabled");
                     Button.goCallback(groupId);
```
im not sure if it made its way into the FFA, but one of the things we've been talking about for a while with Spike v3 is the ability for button pushers to see the current the alert message and edit it before before sending out, when they push the button, bring it in as another step in the confirmation process. might be worth asking bob and shaniqua if we want this in this push as well.

andrewjgremmo replied:
Not in FFA but I can touch base with them about this in our next standup.

kidatsy commented on src/SpikeTeam/UserBundle/Controller/AdminController.php:
```diff
@@ -33,29 +34,25 @@ public function adminAllAction(Request $request)
         if ($form->isValid()) {
             $newAdmin->setPlainPassword($newAdmin->getPassword());
-            $newAdmin->addRole('ROLE_ADMIN');
-            $newAdmin->setEnabled(true);
```
not having this in place makes it so that newly added users (in this case a captain) aren't automatically enabled, in the FOS User sense, and we're not currently exposing that in the admin edit UI, so there's not way to un-disable new accounts (tried logging in after creating a captain here)
<img width="349" alt="screen shot 2016-02-25 at 12 37 49 pm" src="https://cloud.githubusercontent.com/assets/2397451/13328440/a391d9c6-dbbc-11e5-8009-dfdf95b7c101.png">

kidatsy commented on app/config/security.yml:
```diff
@@ -37,7 +38,7 @@ security:
         - { path: ^/api/v2, roles: IS_AUTHENTICATED_ANONYMOUSLY, methods: POST }
-        - { path: ^/, roles: ROLE_ADMIN, requires_channel: https }
+        - { path: ^/, roles: ROLE_CAPTAIN, requires_channel: https }
+        - { path: ^/admin|^/settings, roles: ROLE_ADMIN, requires_channel: https }
```
on login, this is what i see as a captain. do we want captains to see the button page, even if they can't hit button themselves, or only limit to seeing the spike team?
<img width="1184" alt="screen shot 2016-02-25 at 12 39 14 pm" src="https://cloud.githubusercontent.com/assets/2397451/13328511/03d7e5be-dbbd-11e5-8a31-0d9187627dbd.png">

kidatsy replied:
also, captain is still able to access /admin and /settings, its just not in the nav menu. can we lock them out of those routes here?

andrewjgremmo replied:
Yeah I have an updated security.yml file I just haven't committed it because I have the protocols changed so I can run locally.  Same goes for /signup route

andrewjgremmo replied:
And yeah I kept the view of the button because I think even if they don't have the ability to hit it, they should still be able to see when the last time it was hit was.  Shaniqua and Bob are on board with that.

andrewjgremmo commented:
@kidatsy Incorporated most of your feedback.

I left a few things for my next standup with Shannon and Bob like Signup Success page and Button Message stuff.  Also fixed absolute URL generation for Twilio callbacks (you can just pass `true` as the third argument as long as you're running in the context of the web i.e. not a CLI Command, etc) to not require an explicit definition in parameters but for my testing purposes I moved that to a different branch that I'll merge into this when I'm done testing/demoing Twilio for Shaniqua and Bob.

andrewjgremmo commented:
@kidatsy Re: SSL stuff, sorry, that was just for testing.  I had a separate branch for the prod version which will generate an absolute URL instead of building the URL.  Just merged in.

kidatsy commented on web/scss/partials/_variables.scss:
```diff
@@ -1,4 +1,4 @@
-$body-size: 16px;
+$body-size: 14px;
```
good idea. might make sense to make it even smaller..

indspenceable commented on src/SpikeTeam/AlertBundle/Services/NotificationService.php:
```diff
@@ -0,0 +1,86 @@
+
+        $this->settingRepo = $this->em->getRepository('SpikeTeamSettingBundle:Setting');
+        // Pulling CTL Twilio credentials from settings in db
+        $this->sid = $this->settingRepo->findOneByName('twilio_sid')->getSetting();
```
I don't really care, but I think the typical pattern is juuuust set variables in the constructor and try to avoid querying the DB at all.

andrewjgremmo replied:
Good point, I can see how this could go wrong.

indspenceable commented:
Looks fine. :+1:
